File: Solutions/0856/0856.c
/***************************************************************************************************
 * Copyright: washing
 * FileName: 0856.c
 * Author: washing
 * Version: 1.0
 * Date: 2022-10-9 16:17:45
***************************************************************************************************/


int scoreOfParentheses(char * s){
    int * l = (int *)malloc(sizeof(int) * strlen(s));
    int idx = 0, i = 0;
    for (; i<strlen(s); i++) l[i] = -1;
    i = 0;
    for (;i<strlen(s);i++) {
        if (s[i] == '(') l[idx++] = 0;
        else {
            if (l[idx-1] == 0) l[idx-1] = 1;
            else {
                int sum = 0;
                for (; idx>0;) {
                    idx--;
                    if (l[idx] == 0) break;
                    sum += l[idx];
                }
                l[idx++] = sum*2;
            }
        }
    }
    int ret = 0;
    for (int k=0; k<idx; k++) ret += l[k];
    return ret;
}

File: Solutions/0856/0856.py
# -*- coding:utf-8 -*-
# Author: washing
# DateTime: 2022/10/09 15:18
# File: 0856.py
# Desc: 


class Solution:
    def scoreOfParentheses(self, s: str) -> int:
        l = []
        for i in s:
            if i == "(": l.append(i)
            else:
                if l[-1] == '(': l[-1] = 1
                else:
                    idx = len(l) - 1
                    while idx >= 0:
                        if l[idx] == '(': break
                        idx -= 1
                    new = sum(l[idx+1:]) * 2
                    l = l[:idx]
                    l.append(new)
        return sum(l)
